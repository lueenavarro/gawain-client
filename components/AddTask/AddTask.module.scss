.add-task {
  border: 0;
  width: 100%;
  height: 30px;
  padding: 0 20px;
  margin-top: 1.5px;
  background-color: transparent;

  &:focus {
    outline: 0;
  }
}
