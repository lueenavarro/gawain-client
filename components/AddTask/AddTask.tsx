import React, { useState } from "react";
import PropTypes from "prop-types";

import { CharCodes } from "../../constants";
import styles from "./AddTask.module.scss";

interface AddTaskProps {
  onAddTask: Function;
}

const AddTask = ({ onAddTask }: AddTaskProps) => {
  const [value, setValue] = useState("");

  const handleChange = (event) => {
    setValue(event.target.value);
  };

  const handleSubmit = () => {
    onAddTask(value, () => setValue(""));
  };

  const handleBlur = () => {
    if (value) {
      handleSubmit();
    }
  };

  const handleKeypress = (event) => {
    if (value && event.charCode === CharCodes.ENTER) {
      handleSubmit();
    }
  };

  return (
    <input
      className={styles["add-task"]}
      value={value}
      onChange={handleChange}
      onBlur={handleBlur}
      onKeyPress={handleKeypress}
      data-testid="addTaskInput"
    />
  );
};

AddTask.propTypes = {
  onAddTask: PropTypes.func.isRequired,
};

export default AddTask;
