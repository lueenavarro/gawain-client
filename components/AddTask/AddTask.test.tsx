import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import userEvent from "@testing-library/user-event";

import { cleanup, render } from "test/react";
import AddTask from "./AddTask";

const mockOnAddTask = jest.fn((_value, _callback) => {});

beforeEach(() => {
  mockOnAddTask.mockClear();
  cleanup();
});

test("render without crashing", () => {
  const div = document.createElement("div");
  ReactDOM.render(<AddTask onAddTask={mockOnAddTask} />, div);
});

test("call onAddTask when input is blurred", () => {
  const { getByTestId } = render(<AddTask onAddTask={mockOnAddTask} />);
  userEvent.type(getByTestId("addTaskInput"), "Go To Market");
  expect(mockOnAddTask).not.toBeCalled();
  getByTestId("addTaskInput").blur();
  expect(mockOnAddTask).toBeCalledWith("Go To Market", expect.anything());
});

test("don't call onAddTask when input is blurred but empty", () => {
  const { getByTestId } = render(<AddTask onAddTask={mockOnAddTask} />);
  userEvent.type(getByTestId("addTaskInput"), "");
  getByTestId("addTaskInput").blur();
  expect(mockOnAddTask).not.toBeCalled();
});

test("call onAddTask when enter key is pressed", () => {
  const { getByTestId } = render(<AddTask onAddTask={mockOnAddTask} />);
  userEvent.type(getByTestId("addTaskInput"), "Go To Market{enter}");
  expect(mockOnAddTask).toBeCalledWith("Go To Market", expect.anything());
});

test("don't call onAddTask when enter key is pressed but input is empty", () => {
  const { getByTestId } = render(<AddTask onAddTask={mockOnAddTask} />);
  userEvent.type(getByTestId("addTaskInput"), "{enter}");
  expect(mockOnAddTask).not.toBeCalled();
});

test("snapshot matches", () => {
  const tree = renderer.create(<AddTask onAddTask={mockOnAddTask} />).toJSON();
  expect(tree).toMatchSnapshot();
});
