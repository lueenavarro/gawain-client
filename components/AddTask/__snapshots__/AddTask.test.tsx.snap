// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot matches 1`] = `
<input
  className="add-task"
  data-testid="addTaskInput"
  onBlur={[Function]}
  onChange={[Function]}
  onKeyPress={[Function]}
  value=""
/>
`;
